doc/kernel/usermode/kernelobjects.rst

```diff
@@ -0,0 +1,223 @@
+  untrusted or exploited code will then be able to access the object, use
+  this API with caution.
+
+* If a thread was declared statically with :c:macro:`K_THREAD_DEFINE()`,
+  then the :c:macro:`K_THREAD_ACCESS_GRANT()` may be used to grant that thread
+  access to a set of kernel objects at boot time.
+
+Once a thread has been granted access to an object, such access may be
+removed with the :c:func:`k_object_access_revoke()` API. User threads using
+this API must have permission on both the object in question, and the thread
+object that is having access revoked.
+
+Initialization State
+====================
+
+Most operations on kernel objects will fail if the object is considered to be
+in an unitialized state. The appropriate init function for the object must
+be performed first.
+
+Some objects will be implicitly initialized at boot:
+
+* Kernel objects that were declared with static initialization macros
+  (such as :c:macro:`K_SEM_DEFINE` for semaphores) will be in an initialized
+  state at build time.
+
+* Device driver objects are considered initilaized after their init function
+  is run by the kernel early in the boot process.
+
+If a kernel object is intitialized with a private static initailizer, the
+object must have :c:func:`_k_object_init()` on it at some point by a supervisor
+thread, otherwise the kernel will consider the object uninitialized if accessed
+by a user thread. This is very uncommon, typically only for kernel objects that
+are embedded within some larger struct and initialized statically.
+
+.. code-block:: c
+
+    struct foo {
+        struct k_sem sem;
+        ...
+    };
+
+    __kernel struct foo my_foo = {
+        .sem = _K_SEM_INITIALIZER(my_foo.sem, 0, 1),
+        ...
+    };
+
+    ...
+    _k_object_init(&my_foo.sem);
+    ...
+
+
+Creating New Kernel Object Types
+================================
+
+When implementing new kernel features or driver subsystems, it may be necessary
+to define some new kernel object types. There are different steps needed for
+core kernel objects and new driver subsystems.
+
+Creating New Core Kernel Objects
+--------------------------------
+
+* In scripts/gen_kobject_list.py, add the name of the struct to the
+  :py:data:`kobjects` list.
+* The name of the enumerated type is derived from the name of the struct.
+  Take the name of the struct, remove the first two characters, convert to
+  uppercase, and prepend ``K_OBJ_`` to it. Add the enum to
+  :cpp:enum:`k_objects` in include/kernel.h
+* Add a string representation for the enum to the
+  :c:func:`otype_to_str()` function in kernel/userspace.c
+
+Instances of the new struct should now be tracked.
+
+Creating New Driver Subsystem Kernel Objects
+--------------------------------------------
+
+All driver instances are :c:type:`struct device`. They are differentiated by
+what API struct they are set to.
+
+* In scripts/gen_kobject_list.py, add the name of the API struct for the
```

Comment:
nice to mark scripts with double back-ticks:

     ``scripts/gen_kobject_list.py``